### PR TITLE
add image cache framework

### PR DESCRIPTION
**Pull request by neildsouth**
### Description

AC-1240 add image caching 

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New unit tests added to cover the changes.
- [ ] New e2e tests added to cover the changes. (If it has been decided these will not be added with the PR, a seperate ticket **must** be created and linked in the ticket below before merge)
- [ ] All tests passed locally.

### CODE REVIEW

File: src/shared/kubernetes/kuberenetes-client.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            //
+            // Note that this could fail if the spec refers to a custom resource. For custom resources you may need
+            // to specify a different patch merge strategy in the content-type header.
+            //
+            // See: https://github.com/kubernetes/kubernetes/issues/97423
+            const response = await client.patch(spec);
+            created.push(response.body);
+        } catch (e) {
+            // we did not get the resource, so it does not exist, so create it
+            const response = await client.create(spec);
```
Comment (mehranrashid) [inline src/shared/kubernetes/kuberenetes-client.ts:39]: We could add a log here so it is easier to follow in the environments to say that the resource doesn't exist so attempting to create it? Same thing for when it does exist

File: src/shared/kubernetes/kuberenetes-client.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+export async function getPods(namespace: string){
+    const kc = new k8s.KubeConfig();
+    kc.loadFromDefault();
+    const client = kc.makeApiClient(k8s.CoreV1Api);
+    client.listNamespacedPod(namespace).then((res) => {
+        console.log(res.body);
+    },
+        (err) => {
+            console.error('Error!: ' + err);
```
Comment (mehranrashid) [inline src/shared/kubernetes/kuberenetes-client.ts:54]: Are these console logs needed?

File: src/shared/kubernetes/kuberenetes-client.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            }
+        },
+        spec: {
+            cacheSpec: [
+                { images: imageNames }
+            ]
+        }
+    };
+    const manifest = yaml.dump(yml);
+    console.log( `manifest for image cache ${manifest}` );
```
Comment (mehranrashid) [inline src/shared/kubernetes/kuberenetes-client.ts:78]: Could we change this into a log line please?

File: src/shared/kubernetes/kuberenetes-client.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            // try to get the resource, if it does not exist an error will be thrown and we will end up in the catch
+            // block.
+            await client.read(kubeHeader);
+            // we got the resource, so it exists, so patch it
+            //
+            // Note that this could fail if the spec refers to a custom resource. For custom resources you may need
+            // to specify a different patch merge strategy in the content-type header.
+            //
+            // See: https://github.com/kubernetes/kubernetes/issues/97423
+            logger.debug( `ImageCache with name ${name} exists so trying to patch` );
```
Comment (mehranrashid) [inline src/shared/kubernetes/kuberenetes-client.ts:37]: `this.logger.debug`

**Review by DanielNuttall-AD — APPROVED** [review]

**Review by mehranrashid — APPROVED** [review]

**Pull request merged**